toolkit/tools/pkg/imagecustomizerlib/shrinkfilesystems.go

```diff
@@ -0,0 +1,151 @@
+package imagecustomizerlib
+
+import (
+	"fmt"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/imagegen/diskutils"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/logger"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/shell"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+func shrinkFilesystems(imageLoopDevice string, outputImageFile string) error {
+	logger.Log.Infof("Shrinking filesystems")
+
+	// Get partition info
+	diskPartitions, err := diskutils.GetDiskPartitions(imageLoopDevice)
+	if err != nil {
+		return err
+	}
+
+	// Get the start sectors of all partitions
+	matchStarts, err := getStartSectors(imageLoopDevice, len(diskPartitions)-1)
+	if err != nil {
+		return err
+	}
+
+	for partitionNum := 0; partitionNum < len(diskPartitions); partitionNum++ {
+		if diskPartitions[partitionNum].Type == "part" {
+			fstype := diskPartitions[partitionNum].FileSystemType
+			// Currently only support ext2, ext3, ext4 filesystem types
+			if fstype != "ext2" && fstype != "ext3" && fstype != "ext4" {
+				continue
+			}
+
+			partitionLoopDevice := diskPartitions[partitionNum].Path
+
+			// Check the file system with e2fsck
+			err := shell.ExecuteLive(true /*squashErrors*/, "sudo", "e2fsck", "-fy", partitionLoopDevice)
+			if err != nil {
+				return fmt.Errorf("failed to check %s with e2fsck:\n%w", partitionLoopDevice, err)
+			}
+
+			// Shrink the file system with resize2fs -M
+			stdout, stderr, err := shell.Execute("sudo", "resize2fs", "-M", partitionLoopDevice)
+			if err != nil {
+				return fmt.Errorf("failed to resize %s with resize2fs:\n%v", partitionLoopDevice, stderr)
+			}
+
+			// Find the new partition end value
+			end, err := getNewPartitionEndInSectors(stdout, matchStarts[partitionNum-1][1], imageLoopDevice)
+			if err != nil {
+				return err
+			}
+
+			// Resize the partition with parted resizepart
+			_, stderr, err = shell.ExecuteWithStdin("yes" /*stdin*/, "sudo", "parted", "---pretend-input-tty", imageLoopDevice, "resizepart", strconv.Itoa(partitionNum), end)
+			if err != nil {
+				return fmt.Errorf("failed to resizepart %s with parted:\n%v", partitionLoopDevice, stderr)
+			}
+
+			// Re-read the partition table
+			err = shell.ExecuteLive(true, "flock", "--timeout", "5", imageLoopDevice, "partprobe", "-s", imageLoopDevice)
+			if err != nil {
+				return err
```

Comment:
would it make sense to wrap all errors like the `parted` error handling to give more context?

Reply:
enhanced the errors